File: binary-trees/max-path-sum.js
// Leetcode 124. Binary Tree Maximum Path Sum
// Given a non-empty binary tree, find the maximum path sum.
//

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */

const maxPathSum = (root) => {
    const maxSumHelper = (root) => {
        if (!root) return 0;

        const leftSum = maxSumHelper(root.left);
        const rightSum = maxSumHelper(root.right);
        const result = Math.max(leftSum, rightSum, 0) + root.val;
        maxSum = Math.max(maxSum, result, leftSum + rightSum + root.val);
        return result;
    };

    let maxSum = -Infinity
    maxSumHelper(root);
    return maxSum;
};

File: leetcode/binary-trees/max-path-sum.rb
# Leetcode 124. Binary Tree Maximum Path Sum
# Given a non-empty binary tree, find the maximum path sum.
#

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

def cur_max_sum(root)
  return 0 unless root
  left_cur_max = cur_max_sum(root.left)
  right_cur_max = cur_max_sum(root.right)
  cur_max = [left_cur_max, right_cur_max, left_cur_max + right_cur_max, 0].max + root.val
  $maximum = cur_max if cur_max > $maximum
  [left_cur_max, right_cur_max, 0].max + root.val
end

# @param {TreeNode} root
# @return {Integer}
def max_path_sum(root)
  $maximum = -1.fdiv(0)
  cur_max_sum(root)
  $maximum
end
